obfuscation.go

```diff
@@ -0,0 +1,188 @@
+// the shared secret.
+// In context of Lightning Network the obfuscated data is usually a failure
+// which will be propagated back to payment sender, and obfuscated by the
+// forwarding nodes.
+type OnionObfuscator struct {
+	sharedSecret [sha256.Size]byte
+}
+
+// NewOnionObfuscator creates new instance of onion obfuscator.
+func NewOnionObfuscator(router *Router, ephemeralKey *btcec.PublicKey) (*OnionObfuscator,
+	error) {
+
+	sharedSecret, err := router.generateSharedSecret(ephemeralKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OnionObfuscator{
+		sharedSecret: sharedSecret,
+	}, nil
+}
+
+// Obfuscate is used to make data obfuscation.
+// In context of Lightning Network is either used by the nodes in order to
+// make initial obfuscation with the creation of the hmac or by the forwarding
+// nodes for backward failure obfuscation of the onion failure blob. By
+// obfuscating the onion failure on every node in the path we are adding
+// additional step of the security and barrier for malware nodes to retrieve
+// valuable information. The reason for using onion obfuscation is to not give
+// away to the nodes in the payment path the information about the exact failure
+// and its origin.
+func (o *OnionObfuscator) Obfuscate(initial bool, data []byte) []byte {
+	if initial {
+		umKey := generateKey("um", o.sharedSecret)
+		hash := hmac.New(sha256.New, umKey[:])
+		hash.Write(data)
+		h := hash.Sum(nil)
+		data = append(h, data...)
+	}
+
+	return onionObfuscation(o.sharedSecret, data)
+}
+
+// Decode initializes the obfuscator from the byte stream.
+func (o *OnionObfuscator) Decode(r io.Reader) error {
+	_, err := r.Read(o.sharedSecret[:])
+	return err
+}
+
+// Encode writes converted obfuscator in the byte stream.
+func (o *OnionObfuscator) Encode(w io.Writer) error {
+	_, err := w.Write(o.sharedSecret[:])
+	return err
+}
+
+// OnionDeobfuscator represents the serializable object which encapsulate the
+// all necessary data to properly de-obfuscate previously obfuscated data.
+// In context of Lightning Network the data which have to be deobfuscated
+// usually is onion failure.
+type OnionDeobfuscator struct {
+	sessionKey  *btcec.PrivateKey
+	paymentPath []*btcec.PublicKey
+}
+
+// NewOnionDeobfuscator creates new instance of onion deobfuscator.
+func NewOnionDeobfuscator(sessionKey *btcec.PrivateKey,
+	paymentPath []*btcec.PublicKey) (*OnionDeobfuscator, error) {
+	return &OnionDeobfuscator{
+		sessionKey:  sessionKey,
+		paymentPath: paymentPath,
+	}, nil
+}
+
+// Deobfuscate makes data deobfuscation. The onion failure is obfuscated in
+// backward manner, starting from the node where error have occurred, so in
+// order to deobfuscate the error we need get all shared secret and apply
+// obfuscation in reverse order.
+func (o *OnionDeobfuscator) Deobfuscate(obfuscatedData []byte) (*btcec.PublicKey,
+	[]byte, error) {
+	for i, sharedSecret := range generateSharedSecrets(o.paymentPath,
```

Comment:
I may be mis-understanding something, but to me, as currently, written this function is semantically incorrect. Correct me if I'm wrong, but instead we need a double for-loop in order to attempt to decode the packet properly. This is due to the fact that _any_ of the participants on the original route may have originally encrypted the packet. As a result, there are `N^2` different possibilities. 

Let's take a route of `A -> B -> C`. Let's say that `B` was the one who originally created the error. From my reading, this loop would first start with `C`, apply the stream cipher bytes, then proceed to `B` finally. This would result in an invalid final packet as `C`'s stream cipher bytes were erroneously applied.